File: ios/Asthm App/Application/features/firebase/Firebase.swift
import FirebaseAuth
import common

class Firebase {
    
    static var auth: Auth { Auth.auth() }
    
    static func fetchUser(completion: @escaping (common.User?) -> ()) {
        guard let firebaseUser = auth.currentUser else {
            completion(nil)
            return
        }
        
        auth.currentUser?.getIDToken { token, _ in
            if let token = token {
                completion(firebaseUser.toUser(token))
            } else {
                completion(nil)
            }
        }
    }
}

fileprivate extension FirebaseAuth.User {
    
    func toUser(_ token: String) -> common.User {
        return common.User(
            id: uid,
            token: token,
            photoUrl: photoURL?.absoluteString,
            name: displayName,
            email: email,
            isEmailVerified: isEmailVerified,
            phoneNumber: phoneNumber
        )
    }
}

File: app/src/main/java/com/asthmapp/android/firebase/Firebase.kt
package com.asthmapp.android.firebase

import com.asthmapp.model.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser

object Firebase {
    val auth = FirebaseAuth.getInstance()

    fun fetchUser(callback: (User?) -> Unit) {
        val firebaseUser = auth.currentUser ?: run {
            callback(null)
            return@fetchUser
        }
        firebaseUser.getIdToken(false).addOnCompleteListener { task ->
            callback(task.result?.token?.let { token ->
                firebaseUser.toUser(token)
            })
        }
    }

    private fun FirebaseUser.toUser(token: String) = User(
            id = uid,
            name = displayName,
            email = email,
            isEmailVerified = isEmailVerified,
            photoUrl = photoUrl?.toString(),
            token = token,
            phoneNumber = phoneNumber
    )
}
